File: src/main/java/xyz/phanta/cbtweaker/template/TemplateManager.java
package xyz.phanta.cbtweaker.template;

import com.google.gson.JsonObject;
import io.github.phantamanta44.libnine.util.helper.JsonUtils9;
import net.minecraft.util.ResourceLocation;
import xyz.phanta.cbtweaker.CbtMod;
import xyz.phanta.cbtweaker.gui.inventory.UiLayout;
import xyz.phanta.cbtweaker.structure.block.StructureBlockMatcher;
import xyz.phanta.cbtweaker.structure.block.StructureBlockMatcherType;
import xyz.phanta.cbtweaker.util.ConfigException;

import java.nio.file.Path;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class TemplateManager {

    private final TemplateRegistry<UiLayout> uiTemplates;
    private final TemplateRegistry<List<StructureBlockMatcher>> structureBlockMatcherTemplates;

    public TemplateManager(Path templateDir) {
        uiTemplates = new TemplateRegistry<>(
                "UI", templateDir.resolve("ui.json"), dto -> UiLayout.fromJson(dto.getAsJsonObject()));
        structureBlockMatcherTemplates = new TemplateRegistry<>(
                "structure block matcher", templateDir.resolve("structure_block_matcher.json"), dto -> {
            if (dto.isJsonArray()) {
                return JsonUtils9.stream(dto.getAsJsonArray())
                        .map(specDto -> resolveStructureBlockMatcher(specDto.getAsJsonObject()))
                        .collect(Collectors.toList());
            } else {
                return Collections.singletonList(resolveStructureBlockMatcher(dto.getAsJsonObject()));
            }
        });
    }

    private static StructureBlockMatcher resolveStructureBlockMatcher(JsonObject specDto) {
        ResourceLocation typeId = new ResourceLocation(specDto.get("type").getAsString());
        StructureBlockMatcherType type = CbtMod.PROXY.getStructureMatchers().lookUpBlockMatcher(typeId);
        if (type == null) {
            throw new ConfigException("Unknown structure block matcher type: " + typeId);
        }
        return type.loadMatcher(specDto);
    }

    public TemplateRegistry<UiLayout> getUiTemplates() {
        return uiTemplates;
    }

    public TemplateRegistry<List<StructureBlockMatcher>> getStructureBlockMatcherTemplates() {
        return structureBlockMatcherTemplates;
    }

    public void loadPreInit() {
        uiTemplates.loadTemplates();
    }

    public void loadInit() {
        structureBlockMatcherTemplates.loadTemplates();
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/template/TemplateManager.kt
package st.evening.mc.cbtweaker.template

import net.minecraft.util.ResourceLocation
import st.evening.mc.cbtweaker.CbTweaker
import st.evening.mc.cbtweaker.gui.inventory.WindowConfig
import st.evening.mc.cbtweaker.structure.block.StructureBlockMatcher
import st.evening.mc.prelude.api.data.ser.SerializationException
import st.evening.mc.prelude.api.data.tjson.JsonPath
import st.evening.mc.prelude.api.data.tjson.TJson
import st.evening.mc.prelude.api.data.tjson.expectIsObject
import st.evening.mc.prelude.api.data.tjson.forEachObject
import st.evening.mc.prelude.api.data.tjson.useStringValue
import java.nio.file.Path

class TemplateManager(templateDir: Path) {
    companion object {
        context(_: JsonPath)
        private fun resolveStructureBlockMatcher(dto: TJson.Object): StructureBlockMatcher {
            val type = dto.useStringValue("type") {
                CbTweaker.defns.structureBlockMatchers[ResourceLocation(it)]
                    ?: throw SerializationException.withPath("Unknown structure block matcher type: $it")
            }
            return type.loadMatcher(dto)
        }
    }

    val windowTemplates: TemplateRegistry<WindowConfig> = TemplateRegistry(
        "window",
        templateDir.resolve("window.tjson")
    ) { WindowConfig.load(it.expectIsObject()) }

    val structureBlockMatcherTemplates: TemplateRegistry<List<StructureBlockMatcher>> = TemplateRegistry(
        "structure block matcher",
        templateDir.resolve("structure_block_matcher.tjson")
    ) { dto ->
        if (dto is TJson.Array) {
            buildList {
                dto.forEachObject {
                    add(resolveStructureBlockMatcher(it))
                }
            }
        } else {
            listOf(resolveStructureBlockMatcher(dto.expectIsObject()))
        }
    }

    fun loadPreInit() {
        windowTemplates.apply {
            if (CbTweaker.config.builtIns.loadBuiltInWindowTemplates) {
                register("generic_small", WindowConfig.GENERIC_SMALL)
                register("generic_medium", WindowConfig.GENERIC_MEDIUM)
                register("generic_large", WindowConfig.GENERIC_LARGE)
            }
            loadTemplates()
        }
    }

    fun loadInit() {
        structureBlockMatcherTemplates.loadTemplates()
    }
}
